### ISSUE TITLE
two publisher on same instance kill the node process if the rabbitmq server shutsdown 

### ISSUE THREAD

**Issue by mxab** [rudijs/amqp.node-rpc-factory#10]
Hi,
I'm running two publisher with different queues names

```
I20160725-23:57:05.887(2)? Publisher: Unexpected amqplib connection error handled by domain:Error: read ECONNRESET
I20160725-23:57:05.887(2)?     at errnoException (net.js:905:11)
I20160725-23:57:05.888(2)?     at TCP.onread (net.js:559:19)
I20160725-23:57:05.888(2)? Publisher: Unexpected amqplib connection error handled by domain:Error: read ECONNRESET
I20160725-23:57:05.889(2)?     at errnoException (net.js:905:11)
I20160725-23:57:05.889(2)?     at TCP.onread (net.js:559:19)
W20160725-23:57:05.890(2)? (STDERR) 
W20160725-23:57:05.890(2)? (STDERR) TypeError: Cannot set property 'domain' of null
W20160725-23:57:05.890(2)? (STDERR)     at Domain.remove (domain.js:116:13)
W20160725-23:57:05.890(2)? (STDERR)     at Object.<anonymous> (/Users/bruchmann/git/pt/app/node_modules/amqp-rpc-factory/lib/rpc-publisher-factory.js:48:28)
W20160725-23:57:05.890(2)? (STDERR)     at Domain.emit (events.js:117:20)
W20160725-23:57:05.890(2)? (STDERR)     at ChannelModel.emit (events.js:70:21)
W20160725-23:57:05.891(2)? (STDERR)     at Connection.emit (events.js:95:17)
W20160725-23:57:05.891(2)? (STDERR)     at Connection.C.onSocketError (/Users/bruchmann/git/pt/app/node_modules/amqplib/lib/connection.js:351:10)
W20160725-23:57:05.891(2)? (STDERR)     at Socket.emit (events.js:117:20)
W20160725-23:57:05.891(2)? (STDERR)     at net.js:441:14
W20160725-23:57:05.891(2)? (STDERR)     at process._tickDomainCallback (node.js:502:13)

```

from how I understand code this may happen as the publisherDomain is "static" and both publisher receive the error callback as they are in the same domain.
now one of the two publisher may not have published anything so far and therefore the `this.currentConnection` is still null, which results in an error.

I guess on thing to fix it would be to do this in the error handling:

```
this.currentConnection && this.publisherDomain.remove(this.currentConnection);
```

But actually I think the domain should not be shared via initing it directly on the prototype but instead in the init method as you may have two instances with different rabbit server/urls and therefore the error events should not be shared.

Any ideas?

**Comment by mxab** [comment]
@rudijs any thoughts?

**Comment by rudijs** [comment]
Hey @mxab 

I see, interesting.

Can you provide some more details or example code on how I can re-produce this error?

Your using the two publishers from the command line or from within a http server?

I'd like to be able to recreate this before implementing an code changes.

Thanks for you getting this started and creating a pull request, but it does bunch several of the unit tests.

So I think the best course of action is to first re-create the issue, then build on that.

Please let me know when you have time.

Thanks!

**Comment by mxab** [comment]
Hi,
currently I'm using it inside a meteor application, it looks a bit like this:

```
var foo = publisher.create({
    standalone: false,
    url: Meteor.settings.amqp.url,
    queue: "foo_queue",
    replyTimeOutInterval: 1000 * 60 * 3//3 min
})

var bar = publisher.create({
    standalone: false,
    url: Meteor.settings.amqp.url,
    queue: "bar_queue",
    replyTimeOutInterval: 1000 * 15//15secs
})
```

I think both publisher need to be used a least once

```

foo.publish(new Buffer(JSON.stringify({
  dataX...
}))).then(...

bar.publish(new Buffer(JSON.stringify({
  dataY...
}))).then(...

```

When I kill rabbitmq server, the application crashes

I can also try to make real setup

**Comment by rudijs** [comment]
@mxab I see.

So the summary for now is:

The original scope for my needs was a single publisher for both command line and express/koa http servers only.

More than one publisher would be good, but I had no need for that at this time.

I haven't used Meteor apps at all before, this is not ideal as I'd like to re-create the exact issue.

You can see in the examples there's commandline, express and koa. Ideally would be a Meteor example as well.

So the bad news is:

I'm not I'm keen to focus on this.

I'm not using this module at all in production and busy with full time work, so not motivated to spend time on it.

I'm also using node v6 now, so if I was to add functionality I'd use more generators and less promise chains - this means it's more a rewrite than just adding a feature.

So in summary:

At this time, not going to work on this :-(
